### ISSUE TITLE
Electron application menus are being swallowed

### ISSUE THREAD

**Issue by flexiondotorg** [rilian-la-te/vala-panel-appmenu#196]
I've observed this on Ubuntu MATE 18.04 running `unity-gtk-module` (Ubuntu version), MATE 1.20 (and MATE 1.18 with global menu patches) and `vala-panel-appmenu` 0.6.1.

I have noticed that if `appmenu-registrar` is running, regardless of whether the `mate-applet-appmenu` is present in a panel, any Electron applications that are started are having their menus swallowed. Menus are missing from the application regardless of whether `gtk-shell-shows-menubar` in `org.mate.interface` is set True or False. At the time of writing Electron is still linked against GTK2+. 

This issue was not present in `vala-panel-appmenu` 0.5.3.

**Comment by rilian-la-te** [comment]
You threat it as an issue?

We with @lestcape consider it a feature, because in Wayland we does not have way to share ours XSettings to GTK.

**Comment by rilian-la-te** [comment]
If you know alternative way to do Wayland Global Menu sharing - please post it here.

**Comment by flexiondotorg** [comment]
I'm not sure I understand your reply. What do you consider a feature?

**Comment by rilian-la-te** [comment]
Now you do not need to set `gtk-shell-shows-menubar` or any property - just run registrar, and menus will be gone. Or it is gone completely, not shown on panel?

**Comment by rilian-la-te** [comment]
Ok, sorry for my bad English, I will try to explain, why menus swallow when registrar is running.
1. If it is a Qt-based application - it is Qt behaviour. If it sees com.canonical.AppMenu.Registrar in DBus, it swallow a menu and send to Registrar. I was do nothing with it.
2. If you have a dbusmenu-gtk based application, it does the same as Qt application. I was do nothing with it.
3. If you have a regular GTK application, I implemented in appmenu-gtk-module same behaviour.

Electron, AFAIK, is dbusmenu-glib based. So, I was not alternated its behaviour.
I consider it as a feature request to alternate it.

**Comment by rilian-la-te** [comment]
I was mentioned it as problems of HUD-only environment when registrar is running permanently.
Now you can add standalone configure option on your HUD. When it is activated, HUD will quit a Registrar when hidden (it has a DBus method for this)

**Comment by rilian-la-te** [comment]
@flexiondotorg do you read it?

**Comment by lestcape** [comment]
Probably, could be a good idea add a gsettings as an extension settings to control the registrar behaviour. In that way, will exist a way to disable/enable the swallowed of the window menu by affecting the registrar status without affecting the extension status as a whole. But to be honest, i'm with the idea of if you have the extension enable is because you want to see a global menu, otherwise have not sense to have enable an extension for a global menu in your desktop. If the extension is not an extension an is a default componet of your desktop, well you will really need another setting, but if it's an extension, enable/disable the extension mean enable/disable the menu behavior in my opinion. So, i'm with the Qt idea of how this work and not with the idea of to have a lot of settings.

**Comment by lestcape** [comment]
The problem of to have add a gsettings as an extension settings to control the registrar behaviour, will be this issue: https://bugs.launchpad.net/appmenu-qt5/+bug/1740171

**Comment by rilian-la-te** [comment]
This is not a good bug. But this is not my bug for now.
Keep this as a tracker.

**Comment by rilian-la-te** [comment]
Or you want me to add setting to registrar?
If yes, what such setting should do?

**Comment by flexiondotorg** [comment]
Let me explain the issue I am seeing more clearly.

## Without appmenu panel applet in a MATE panel

  * I am using `unity-gtk-module` on Ubuntu MATE.
  * `mate-hud` activates `appmenu-registrar` via Dbus. This is good, and what I requested.
  * The appmenu panel applet is _not_ in any panel.
  * Launch a GTK2 or a LibreOffice application and it _does not have menus_.
  * Launch a GTK3 application and it _has menus_.

## With appmenu panel applet in a MATE panel

  * I am using `unity-gtk-module` on Ubuntu MATE.
  * The appmenu panel applet _is_ in a panel.
  * Launch a GTK2 or LibreOffice application and menus are _only_ displayed in the appmenu panel applet.
  * Launch a GTK3 application and menus are displayed _both locally_ in the application _and in the appmenu panel applet_, unless I toggle `gtk-shell-shows-menubar` to True in `org.mate.interface`.

## My request

Menus should not be swallowed by `appmenu-registrar` alone.

  * If `appmenu-registrar` is running but the appmenu panel applet is not, then menus should not be swallowed and should only be presented locally in all applications.
  * If the appmenu panel applet is running, then menus should automatically be presented in the global menu only. Rather than what happens now where some applications present menus locally and globally unless `gtk-shell-shows-menubar` is True.
  * Removing the appmenu panel applet from the panel should automatically restore menus locally to all applications.

I hope that better explains the inconsistency I'm seeing?

**Comment by flexiondotorg** [comment]
I have also tested this with `appmenu-gtk-module` 0.6.1 and the results are the same. I've updated the first comment.

**Comment by rilian-la-te** [comment]
1. Can you test it with master and post behaviour here?
Master is very different from 0.6.1 now, I am prepare a next major release (0.7.0) with bold appname and compact mode. And new dbusmenu engine.

**Comment by rilian-la-te** [comment]
I will explain it futher this evening (about toolkits and why I cannot control menu swallowing on Qt applications and Electron applications)

**Comment by rilian-la-te** [comment]
And is your GTK build with GDK_WINDOWING_WAYLAND?

**Comment by flexiondotorg** [comment]
I've followed your advice and updated `mate-hud` to terminate `appmenu-registrar` after each query, so long as `appmenu-mate` (the panel applet) is not running.

I'll check the Ubuntu build options for GTK and report back. I'll test a build from master as soon as I get some time to do so.

**Comment by rilian-la-te** [comment]
Not terminate, but use dbus quit method, it will free everything.
If you just SIGTERM it, it can leaks memory.

And can you give an option to set applet name?
Because we can use ValaPanel or xfce4-panel in MATE too. And these panels have different naming schemes for applet (or do not have applet name at all if applets is in-process. Mate can also have in-process applets, so does not have applet name)

**Comment by rilian-la-te** [comment]
1.  Look to ```com.canonical.AppMenu.Registrar``` protocol.
```Vala
    [DBus (name = "com.canonical.AppMenu.Registrar")]
    public interface OuterRegistrar : DBusProxy
    {
        public signal void window_registered(uint window_id, string service, ObjectPath path);
        public signal void window_unregistered(uint window_id);
        public abstract void register_window(uint window, ObjectPath menu_object_path) throws Error;
        public abstract void unregister_window(uint window) throws Error;
        public abstract void get_menu_for_window(uint window, out string service, out ObjectPath path) throws Error;
        public abstract void get_menus([DBus (signature="a(uso)")] out Variant menus) throws Error;
    }
```

This is no place for register an applet or so. No property, no method. Adding this will not be supported by other AppMenu providers (Unity)
2. For control Qt apps behaviour I need to write a QPA platformtheme, which I do not want to do now.
3. For control Electron apps - I do not know how to control Electron internals. It links with Gtk, but it uses dbusmenu-glib directly (look to [this](https://github.com/electron/electron/blob/master/atom/browser/ui/views/global_menu_bar_x11.cc))
4. I also do not know how to control LibreOffice internals, it uses GMenuModel directly to display a menu AFAIK.


And then how I should implement your request?

**Comment by flexiondotorg** [comment]
@rilian-la-te Thanks for all the useful information. I've been able to mostly workaround the issue I observed by modifying `mate-hud` and mate-tweak` to terminate `appmenu-registrar`. The solution is a proof of concept at the moment, using SIGTERM, but I will switch to DBus quit method as you suggested and make sure panels for other desktop environments are supported.

So, my request now becomes this:

  * When `appmenu-mate` is added to a panel can it automatically set `gtk-shell-shows-menubar` in `org.mate.interface` to _True_?
  * When `appmenu-mate` is removed from a panel can it automatically set `gtk-shell-shows-menubar` in `org.mate.interface` to _False_?

This would make the effect of adding `appmenu-mate` to a panel consistent as most toolkits/frameworks respond to the presence of `appmenu-registrar` automatically but GTK3 applications do not.

**Comment by rilian-la-te** [comment]
1. It should also takes registrar presence into account. If not in master, I will need to fix it in appmenu-gtk-module.

> 
    When appmenu-mate is added to a panel can it automatically set gtk-shell-shows-menubar in org.mate.interface to True?
    When appmenu-mate is removed from a panel can it automatically set gtk-shell-shows-menubar in org.mate.interface to False? Perhaps even send Quit to appmenu-registrar?

This request are MATE-specific, i do not support this, because.
1. If GTK3 applications do not obey registrar presence - it is a bug and should be threated appropriately.
2. I will not quit a Registrar, because you can:
a) have 2 menus at once (on all monitors, for example)
b) have 2 different panels where both have a global menu (so, registrar is required)
c) have non-standard registrar (KWin)

**Comment by rilian-la-te** [comment]
You have a Registrar inside an extension, but my Registrar is standalone application, and I think than user can create may copies of my widget (which is not true for GNOME extension).
So, you are correct for GNOME, but But I think not for rest.

**Comment by lestcape** [comment]
My extension if a fork of a cinnamon extension, where i also kill registrar... In Cinnamon multi-instance are allowed. But i see your point, because yes registrar need to be one per desktop-session, but an extension is not considered as a part of a desktop-session. So, this is really problematic and difficult to decide. For another hand, who say that i can not create multi-instance in gnome. In fact i can run Cinnamon as an extension of gnome (with the support of multiple instance also embedded): https://github.com/lestcape/Classic-Gnome

**Comment by lestcape** [comment]
Well, I think you can not use directly registrar to enable the gtk2 and Gtk3 wayland window menubar if your plan will be not kill the registrar iface, because this have unwanted effects from users. If there are not a client extension that is able to handled the menubar, is for me not good at all remove the menubar from the windows, because user will have not a way to interact with the menubar. This is my point of view.

**Comment by lestcape** [comment]
To create multi-instance for extensions my idea is shared an instance manager between all extensions and just kill the registar when the last instance will be removed and instantiate registar when the first  instance is create. This ofcourse will not allow any other extension to used my registar instance. Like an individual hub extension for example.

**Comment by rilian-la-te** [comment]
@lestcape  please look to [here](https://github.com/rilian-la-te/vala-panel-appmenu/blob/0.5.6/lib/registrar.vala), it was in 0.5 series.

But later I decided than this is not good, because if applet crashed, menus will be gone forever ([Launchpad issue](https://bugs.launchpad.net/appmenu-qt5/+bug/1740171)). So, I made standalone registrar and DBus activation files for it.
And if I will threate it as inner - it will have no difference AFAIK.

Your idea for me sounds as 0.5 series registrar.

**Comment by rilian-la-te** [comment]
> Well, I think you can not use directly registrar to enable the gtk2 and Gtk3 wayland window menubar if your plan will be not kill the registrar iface, because this have unwanted effects from users. If there are not a client extension that is able to handled the menubar, is for me not good at all remove the menubar from the windows, because user will have not a way to interact with the menubar. This is my point of view.

And how we should handle crashed applets then? Menus in Qt applications will gone forever.

**Comment by lestcape** [comment]
Qt have a bug that need to be fixed: https://bugs.launchpad.net/appmenu-qt5/+bug/1740171 because yes if registrar is gone it can not be recovered. After that a crash can be handled removing and adding the extension again.

**Comment by rilian-la-te** [comment]
Ok, I will look for it)

**Comment by rilian-la-te** [comment]
Crash - yes. But Qt menus will not be recovered then.
About my registrar: I can add commandline ref and unref options to it (using g_application_hold and g_application_release). 
And in MenuWidget can call ref and unref for it.

**Comment by rilian-la-te** [comment]
@flexiondotorg what do you think about registrar ref and unref?
But calling it - I need to do it optional and do not catch an error (because if you run another registrar, you cannot unref it, it is OK)

**Comment by lestcape** [comment]
> Crash - yes. But Qt menus will not be recovered then.

Because this  bug yes, can not be recovered: https://bugs.launchpad.net/appmenu-qt5/+bug/1740171 It need to be fixed.

**Comment by lestcape** [comment]
Another way will be include the appmenu-qt5 in your code and we can try to fix it for us. To be honest, i don't think ubuntu will fix anything related with the global menu. They are now friends of the gnome way, with certainly, they have a complete different road map for the menubar now.

**Comment by rilian-la-te** [comment]
I do not have appmenu-qt5) I use qt builtin menubar handling. And I think use appmenu-qt5 is a bad way - user wants qt5ct.

**Comment by lestcape** [comment]
Ok, appmenu-qt5 it's installed by default with unity (so i have that bug). If you don't use appmenu-qt5 then is supposed that if you use qt builtin, this have not the issue or also it have it?

**Comment by rilian-la-te** [comment]
Issue exists in all DEs except KDE, AFAIK.

**Comment by lestcape** [comment]
It is a known and fixed bug on qt4, for the ubuntu implementation, with is supposed was ported to Qt to be the Qt build in:
https://bugs.launchpad.net/appmenu-qt/+bug/838115
I really don't know where is that code on Qt or how it work.

**Comment by rilian-la-te** [comment]
Now registrar gets reference and unreference support (commandline).
So, when I update all applets to ref and unref a registrar - issue will be resolved (AFAIK).
But registrar will swallow a menu anyway.

**Comment by flexiondotorg** [comment]
Should I change something in `mate-hud` to use reference and unreference?

**Comment by rilian-la-te** [comment]
For now, it is experimental, because no applet has support of it.
You can reference a Registrar after starting a hud, and unref it after end. But it will have one more reference (because no applet can ref it)
After applets implemented a reference, a mechanic will be changed to use it.

But it is experimental and I needs your opinion:
1. How it will interfere with dbus-launch?
2. May private dbus protocol will be better (with ignoring errors)?

**Comment by rilian-la-te** [comment]
@flexiondotorg i need to do it before 0.7.0 release. I need to decide, how I should do better, by your opinion:
1. Return to handle GTK2 via GSettings
2. Add ref/unref DBus private proto
3. Add ref/unref commandline
4. Stay as is

**Comment by rilian-la-te** [comment]
Implemented reference and unreference as DBus properties.
Now, if registrar is not referenced, it will exits after 10 seconds.
vala-panel-appmenu reference and unreference registrar by default.

**Comment by nkkollaw** [comment]
I'm not sure what this entry is about, but is it about menus on GTK2 and Electron apps not showing? Because they don't seem to show :-)

**Comment by rilian-la-te** [comment]
Screen. And then I will decide about this.

**Comment by rilian-la-te** [comment]
@flexiondotorg and what about 0.6.90 refcounting registrar?

**Comment by flexiondotorg** [comment]
@rilian-la-te This is something I've been meaning to get back to. How should I use reference and unreference from my HUD? Czn you explain what reference and unreference do exactly?

**Comment by rilian-la-te** [comment]
Just use dbus methods Reference or UnReference on org.valapanel.AppMenu.Registrar. It increases or decreases a refcount of Registrar. If refcount drops to zero, Registrar will exit.

**Comment by flexiondotorg** [comment]
So HUD should register when starting a query and then unregister when finished?

**Comment by rilian-la-te** [comment]
Yes. And if Registrar is not running, it should dbus activate it first.

**Comment by rilian-la-te** [comment]
So, what about this approach?

**Comment by lestcape** [comment]
**IMO:** Nice.

**Comment by rilian-la-te** [comment]
@lestcape thanks.
Waiting to @flexiondotorg decision about this.

**Comment by snippins** [comment]
@rilian-la-te I think you can change your approach just a little bit. When ref count = 0 it appmenu-registar does not need to quit, just does not act as a Registrar, does not swallow menus and stay in memory. This way it could be start once as a service in the session and does not need to keep being open and  close if the user do not use the global menu. This also helps that the use case of Ref and Unref does not need any change whether one use global menu or not.

**Comment by rilian-la-te** [comment]
It is more complex than usual. I will think about it.

**Comment by lestcape** [comment]
@snippins appmenu-gtk-module is observing the registrar and is acting according to the presence of the service. To apply your idea is needed an introduction of another different way to control the unity-gtk-module with is overhead the process in my opinion, but well.

**Comment by rilian-la-te** [comment]
I do not like an approach to having a process when it is not needed and just staying in proc table.

**Comment by flexiondotorg** [comment]
When MATE HUD register/activates appmenu-registrar it immediately swallows menus. This means when the HUD is requested any open applications "dance" as the menus are removed and then reinstated.

I would like the means to delay the menus being removed from the application. I realise this is not a requirement for most, but if a gschema setting could be provided to provide a delay (default 0) I could override it in Ubuntu MATE (approximately 250 to 500 ms) and prevent the menus jumping about when the HUD is requested.

**Comment by rilian-la-te** [comment]
But I cannot control any application than does not use appmenu-gtk-module.
But if I will add a delay to appmenu-gtk-module, then its behaviour will be different than other apps. Do you OK with that?

**Comment by flexiondotorg** [comment]
Yes, I think that would be a significant improvement.

**Comment by rilian-la-te** [comment]
Ok, I will think about it. But it will cause a difference of behaviour for Gtk and Qt applications.

**Comment by flexiondotorg** [comment]
In my testing, Qt and Electron applications are not affected by this issue. Firefox, Thunderbird, LibreOffice, GTK3, wx (such as Audacity) applications and Gimp are.

**Comment by rilian-la-te** [comment]
Firefox and Thunderbird will not be fixed, because it should be fixed in Unity patch.

For other applications - can you get a Qt applications delay and post it here?

**Comment by rilian-la-te** [comment]
And I want a video about this issue (I understand that this issue is HUD-only)

**Comment by lestcape** [comment]
@rilian-la-te yes, it will be better if you find the cause and fixed it... But if not, you also have a delay here:
https://github.com/rilian-la-te/vala-dbusmenu/commit/84daace18ce8327ac1497747fd362727904d5783#diff-2977b68e34896a6e33561c7595876802L232

Probably you can synchronized all delayed?

**Comment by rilian-la-te** [comment]
@lestcape vala-dbusmenu is not used now, Now I use Importer for this task.
And this delay was for Qt applications, not for Gtk.

**Comment by lestcape** [comment]
Yes exactly, because was for QT, and this occurs for GTK, if both have the same delay value they can be synchronized.

> Ok, I will think about it. But it will cause a difference of behaviour for Gtk and Qt applications.

But if is not used now, no way...

**Comment by rilian-la-te** [comment]
@flexiondotorg can you send me a video of an issue?

**Comment by rilian-la-te** [comment]
@flexiondotorg I have added a key to disable swallowing a menus in appmenu-gtk-module. But it will not work on the fly.

**Comment by rilian-la-te** [comment]
If you accept this solution, there will be new release.

**Comment by flexiondotorg** [comment]
@rilian-la-te I will test ASAP.

**Comment by flexiondotorg** [comment]
@rilian-la-te I have applied the "inner menu" patches, enabled the `always-show-inner-menu` key and rebooted.

See this video of what happens when the HUD is activated, which in turn briefly dbus activates `appmenu-registrar`. You will see some applications have their menus swallowed for a moment. In this example, the global menu applet has not been added to any panel.

https://imgur.com/a/EFBnN

I also noticed that when the global menu applet is present in a panel, some application will now have local and global menus, but should have global menus only.

**Comment by rilian-la-te** [comment]
> I also noticed that when the global menu applet is present in a panel, some application will now have local and global menus, but should have global menus only.

If you want a global menu, inner menu key should be disabled.

Can you give me a list of applications where menu is swallowed?
1. Firefox - I cannot control this by appmenu-gtk-module, because it uses XUL and Unity patches works independent from me (it will swallows by it when Registrar is found).
2. LibreOffice - I do not know about this, but this is more realistic.

Another applications not swallow, AFAIS.
Can you give me an exact list of swallowing application, and I then will describe, which application can be fixed, and if can, how, and if me can fix it.

**Comment by rilian-la-te** [comment]
@flexiondotorg how you want to solve a problems with applications which does not use appmenu-gtk-module, like LibreOffice or Firefox (first is homebrew toolkit, and second is XUL)?
